src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java

```diff
@@ -895,8 +896,9 @@ public boolean isComplete() {
         first.setNewHead(head);
         heads.clear();
         heads.put(first.getId(),first);
+    }
 
-        // clean up heap
+    synchronized void cleanUpHeap() {
```

Comment:
`register` call in original bug fix added some objects to a record in this class. `SerializableClassRegistry.release` was getting called on the right loader, but too soon—_before_ the final call to `saveProgram`, which would up adding the loader back to the registry.